BDArmory/Guidances/MissileGuidance.cs

```diff
@@ -147,6 +147,184 @@ public static Vector3 GetAirToGroundTarget(Vector3 targetPosition, Vector3 targe
             return targetPosition + (targetVelocity * leadTime);
         }
 
+        public static Vector3 GetAirToAirLoftTarget(Vector3 targetPosition, Vector3 targetVelocity,
+            Vector3 targetAcceleration, Vessel missileVessel, float targetAlt, float maxAltitude,
+            float rangeFactor, float altComp, float velComp, float loftAngle, float termAngle,
+            float termDist, ref int loftState, out float timeToImpact, out float targetDistance,
+            float minSpeed = 200)
+        {
+
+            Vector3 velDirection = missileVessel.Velocity().normalized;
+
+            targetDistance = Vector3.Distance(targetPosition, missileVessel.transform.position);
+
+            float currSpeed = Mathf.Max((float)missileVessel.srfSpeed, minSpeed);
+            Vector3 currVel = currSpeed * velDirection;
+
+            //Vector3 Rdir = (targetPosition - missileVessel.transform.position).normalized;
+            //float rDot = Vector3.Dot(targetVelocity - currVel, Rdir);
+
+            float leadTime = (float)(1 / ((targetVelocity - currVel).magnitude / targetDistance));
+            //float leadTime = (targetDistance / rDot);
+
+            timeToImpact = leadTime;
+            leadTime = Mathf.Clamp(leadTime, 0f, 8f);
+
+            // If loft is not terminal
+            if ((targetDistance > termDist) && (loftState < 3))
+            {
+                if (BDArmorySettings.DEBUG_MISSILES) Debug.Log("[BDArmory.MissileLauncher]: Lofting");
+
+                // Get up direction
+                Vector3 upDirection = VectorUtils.GetUpDirection(missileVessel.CoM);
+
+                // Use the gun aim-assist logic to determine ballistic angle (assuming no drag)
+                Vector3 bulletRelativePosition, bulletRelativeVelocity, bulletAcceleration, bulletRelativeAcceleration, targetPredictedPosition, bulletDropOffset, lastVelDirection, ballisticTarget, targetHorVel, targetCompVel;
+                
+                var firePosition = missileVessel.transform.position + (currSpeed * velDirection) * Time.fixedDeltaTime; // Bullets are initially placed up to 1 frame ahead (iTime). Not offsetting by part vel gives the correct initial placement.
+                bulletRelativePosition = targetPosition - firePosition;
+                float timeToCPA = timeToImpact; // Rough initial estimate.
+                targetPredictedPosition = AIUtils.PredictPosition(targetPosition, targetVelocity, targetAcceleration, timeToCPA);
+
+                // Velocity Compensation Logic
+                float compMult = Mathf.Clamp(0.5f * (targetDistance - termDist) / termDist, 0f, 1f);
+                Vector3 velDirectionHor = (velDirection - upDirection * Vector3.Dot(velDirection, upDirection)).normalized;
```

Comment:
`Vector3.ProjectOnPlane(velDirection, upDirection).normalized` does this and is probably optimised better.
Similarly for `targetHorVel` (without the `.normalized`).

Reply:
Replaced, however my method should be more efficient as I assume upDirection is normalized unlike ProjectOnPlane, which uses 3 division operations:
`public static Vector3 ProjectOnPlane(Vector3 vector, Vector3 planeNormal)
{
    float sqrMag = Dot(planeNormal, planeNormal);
    if (sqrMag < Mathf.Epsilon)
        return vector;
    else
    {
        var dot = Dot(vector, planeNormal);
        return new Vector3(vector.x - planeNormal.x * dot / sqrMag,
            vector.y - planeNormal.y * dot / sqrMag,
            vector.z - planeNormal.z * dot / sqrMag);
    }
}`

Reply:
Huh, fair enough, yeah, if you know the `planeNormal` is already normalised, then what you've written ought to be faster. I'll benchmark it tomorrow to see what difference it makes. Really, Unity should have used
```csharp
var dotNorm = Dot(vector, planeNormal) / sqrMag;
return new Vector3(vector.x - planeNormal.x * dotNorm,
            vector.y - planeNormal.y * dotNorm,
            vector.z - planeNormal.z * dotNorm);
```
to reduce their number of operations.

Reply:
Updated